Support multiple :name:s with multiple signatures for .. tacn::

jfehrle opened the pull request:
Needed for #17503 (See this [comment](https://github.com/coq/coq/pull/17503#discussion_r1170207061)).

@Zimmi48 Can you review/merge this very short PR soon?

Zimmi48 reviewed (changes requested):
I think that accepting more names than signature lines and silently pointing extra names to the first signature line is a footgun (even more so without any documentation of this behavior). It seems that your suggestion from https://github.com/coq/coq/pull/17503#discussion_r1170207061 illustrates this, since this should have been `:names: lazy cbv weak lazy` (and located on the line below).
Since the use case appears valid, a semantic that would look saner to me would simply be to point all the names to the first signature in case there are more names than signatures. This even leads to a simpler implementation:

```python
            if len(names) < len(sigs):
                ERR = ("Expected at least {} semicolon-separated names, got {}.  " +
                       "Please provide at least one name per signature line.")
                raise self.error(ERR.format(len(names), len(sigs)))
            elif len(names) > len(sigs):
                self._sig_names = { sigs[0]: names }
            else:
                self._sig_names = { sig: [name] for (sig, name) in zip(sigs, names) }
```

Though there might also be valid use cases for providing fewer names than signature lines (in case three signatures start with the same two possible names, but end up with different parsing rules), in which case the code would simply be:

```python
            if len(names) != len(sigs):
                self._sig_names = { sigs[0]: names }
            else:
                self._sig_names = { sig: [name] for (sig, name) in zip(sigs, names) }
```

Unless @cpitclaudel has a better idea.

jfehrle commented:
> I think that accepting more names than signature lines and silently pointing extra names to the first signature line, but not doing so for the middle names, is a footgun (even more so without any documentation of this behavior).

Colorful, but this seems a minor problem.  We're never going to be able to check for every possible doc error.  You even suggest pointing all the names to the first signature.  I'm OK with changing it so all signatures point to the first entry.

> It seems that your suggestion from https://github.com/coq/coq/pull/17503#discussion_r1170207061 illustrates this, since this should have been `:names: lazy cbv weak lazy` (and located on the line below).

Not so much, one has to test such changes.  Hard to make review comments flawless when you're not actually making the changes.  FWIW, I won't mention that you should have said `:names: lazy; cbv; weak lazy` above.

Zimmi48 commented on doc/tools/coqrst/coqdomain.py:
```diff
@@ -274,14 +274,16 @@ def _prepare_names(self):
         else:
             names = [n.strip() for n in names.split(";")]
-            if len(names) != len(sigs):
+            if len(names) < len(sigs):
```
What makes you think I meant something else?

jfehrle replied:
Syntax.  What do you mean by (simplifying) "A > B = 1"?

Zimmi48 replied:
Sorry for the delayed answer. I mean `A > B /\ B = 1`.

cpitclaudel commented:
This looks OK to me, but doesn't sphinx let you add custom index entries with `.. index`?  Maybe that's impler than adding more cases to the already-complicated `:name:` syntax?  Not sure.

jfehrle commented on doc/tools/coqrst/coqdomain.py:
```diff
@@ -274,14 +274,16 @@ def _prepare_names(self):
+                    ERR = ("Expected at least {} semicolon-separated names, got {}.  " +
+                           "Please provide at least one name per signature line.")
                     raise self.error(ERR.format(len(names), len(sigs)))
                 self._sig_names = { sigs[0]: names }
```
It's reachable, but the message needs to be reworded.

jfehrle commented:
Updated.  Hope we can wrap this up very soon.

I simplified the behavior.  No need to change ~100 uses of `:name:` to `:index:`

I removed some unneeded `:name:`s.

I couldn't remember how to add text to `doc/sphinx/README.rst`.  I wanted to add "If ``:name:`` is specified, the number of names must be >= the number of signatures.  The name ``_`` will not appear in indexes."  Do you remember?

cpitclaudel commented:
> I simplified the behavior. No need to change ~100 uses of :name: to :index:

I wasn't proposing that, just to handle this additional new case with index instead of making :name: more powerful. (Or did I miss something?)

Zimmi48 commented:
> I couldn't remember how to add text to `doc/sphinx/README.rst`. I wanted to add "If `:name:` is specified, the number of names must be >= the number of signatures. The name `_` will not appear in indexes." Do you remember?

You need to modify the `README.template.rst` file and run the build to update `README.rst`.

Zimmi48 reviewed (approved):
I'm fine with the current version, but maybe we should have kept the ability of linking to each signature when the number of provided names matches exactly with the number of signatures?

I'm also not completely clear on what the alternative proposed by Clément would entail.

BTW, there is a build error that I'm not sure that I understand.

jfehrle commented:
@cpitclaudel The Coq doc has 7 indexes.  I don't see how `.. index::` would know which one should get the entry.

jfehrle commented:
That's the same message it was getting before, you just had to look at the full log to see it.  I only rebased, no code change, but I think I figured out the cause.

jfehrle commented:
I still can't get this to work after several hours of trying :-(.  Perhaps one of you can figure out how to do this.  It does seem like a useful enhancement if it can be made to work.

cpitclaudel commented:
> @cpitclaudel The Coq doc has 7 indexes. I don't see how .. index:: would know which one should get the entry.

See https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-index, you give it the name of the index and what index entry to use.

cpitclaudel commented:
>  Perhaps one of you can figure out how to do this. 

I would love to help, but I won't have time to write code for this any time soon.  If it's critical, I will try to make time.  But maybe the .. index:: solution is enough? Let me know

jfehrle commented:
> If it's critical, I will try to make time.

Thanks.  The immediate need for this for #17503 has gone away--they'll probably use different syntax that won't need an extra index entry.

> But maybe the .. index:: solution is enough?

I looked at the referenced Sphinx doc twice and closely but I don't see how you specify the name of the index.  Is it something like `:name:`??

cpitclaudel commented:
No, you're right and I was wrong.  `.. index` only inserts entries in the main index...

gares commented:
Not ready for 8.18, postponing to 8.19

SkySkimmer commented:
It seems this PR stalled months ago, I'm removing the milestone.

SkySkimmer commented:
Doesn't work and stalled, reopen if there's progress
